src/utils/url/Url.ts

```diff
@@ -1,12 +1,25 @@
 import { IUrl } from "./IUrl";
+import { Protocol } from "./Protocol";
+import { copyMutateAndWrap } from "./copyMutateAndWrap";
+import { defaultPort } from "./defaultPort";
 
-export const Url = (urlString:string):IUrl => ({
-    get path() { return new URL(urlString).pathname },
-    asString: urlString,
-    setPath(newPath) { 
-        const tempUrl = new URL(urlString);
-        tempUrl.pathname = newPath
-        return Url(tempUrl.toString())
-    },
-    toString: () => urlString,
+export const Url = (nativeURL:URL):IUrl => ({
+    get path() { return nativeURL.pathname },
+    get host() { return nativeURL.host },
+    get protocol() { return nativeURL.protocol.replace(":","") as Protocol },
+    get asString() { return nativeURL.toString() },
+    get toHttps() { return this.setProtocol("https") },
+    get port() { return Number(nativeURL.port) },
+    setPath: (newPath) => copyMutateAndWrap(nativeURL, u => u.pathname = newPath),
+    setProtocol: (newProtocol) => copyMutateAndWrap(nativeURL, u => u.protocol = newProtocol),
+    setHost: (newHost) => copyMutateAndWrap(nativeURL, u => u.host = newHost),
+    toString() { return this.asString },
+    setPort(newPort) {
+        return copyMutateAndWrap(
+            nativeURL, 
+            u => u.port = newPort == defaultPort(this.protocol) 
+                ? "" 
```

Comment:
is there a non-cosmetic reason for setting port to empty string if newPort is the default web port? and if so, would it be better to set it to None/undefined instead

Reply:
It’s a built in type, URL, and it requires a string; setting to undefined yielded a compile error if i recall correctly.

Luckily we can mask this ugliness in our IUrl wrapper around this native api

Reply:
So for example it was adding an explicit :80 or explicit 443 when the load balancer redirected in the unit tests; so I realized I wanted to eliminate that behavior, so I wanted to check, if it is the default port for the given protocol, clear the value, yielding a passing test; but to clear I had to set empty string; else couldn’t compile 🙂

Reply:
So for example it was adding an explicit :80 or explicit 443 when the load balancer redirected in the unit tests; so I realized I wanted to eliminate that behavior, so I wanted to check, if it is the default port for the given protocol, clear the value, yielding a passing test; but to clear I had to set empty string; else couldn’t compile 🙂